From: =?utf-8?q?Fr=C3=A9d=C3=A9ric_Devernay?= <frederic.devernay@m4x.org>
Date: Tue, 27 Aug 2019 21:51:16 -0700
Subject: CImg - make Inpaint abortable

Extracted from https://github.com/NatronGitHub/openfx-misc.git
See File: CImg/Inpaint/inpaint.h.patch

Initial revision: c450e23bba0a
Last revision (CImg: update to 2.7.0) 9db7c40aa0cd5e9
---
 CImg/Inpaint/inpaint.h | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/CImg/Inpaint/inpaint.h b/CImg/Inpaint/inpaint.h
index d7083f9..5c3c892 100644
--- a/CImg/Inpaint/inpaint.h
+++ b/CImg/Inpaint/inpaint.h
@@ -103,6 +103,7 @@ CImg<T>& inpaint_patch(const CImg<t>& mask, const unsigned int patch_size=11,
   // Find (dilated by 2) bounding box for the inpainting mask.
   unsigned int xm0 = _width, ym0 = _height, xm1 = 0, ym1 = 0;
   bool is_mask_found = false;
+  cimg_pragma_openmp(parallel for collapse(2) cimg_openmp_if(mask.width()>=256 && mask.height()>=16))
   cimg_forXY(mask,x,y) if (mask(x,y)) {
     is_mask_found = true;
     if (x<(int)xm0) xm0 = (unsigned int)x;
@@ -139,7 +140,10 @@ CImg<T>& inpaint_patch(const CImg<t>& mask, const unsigned int patch_size=11,
   weights.draw_gaussian((float)p1,(float)p1,patch_size/15.0f,&one)/=patch_size2;
   unsigned int target_index = 0;
 
+  _cimg_abort_init_openmp;
+  cimg_abort_init;
   while (true) {
+    cimg_abort_test;
 
     // Extract mask border points and compute priorities to find target point.
     unsigned int nb_border_points = 0;
@@ -416,6 +420,7 @@ CImg<T>& inpaint_patch(const CImg<t>& mask, const unsigned int patch_size=11,
     // Generate blending scales.
     CImg<T> result = _inpaint_patch_crop(ox,oy,ox + dx - 1,oy + dy - 1,0);
     for (unsigned int blend_iter = 1; blend_iter<=blend_scales; ++blend_iter) {
+      cimg_abort_test;
       const unsigned int
         _blend_width = blend_iter*blend_size/blend_scales,
         blend_width = _blend_width?_blend_width + 1 - (_blend_width%2):0;
@@ -454,19 +459,25 @@ CImg<T>& inpaint_patch(const CImg<t>& mask, const unsigned int patch_size=11,
       }
 
       if (is_blend_outer) {
+        cimg_pragma_openmp(parallel for collapse(2) cimg_openmp_if(blended.width()>=256 && blended.height()>=16))
         cimg_forXY(blended,x,y) if (blend_map(x,y)==blend_iter) {
           const float cum = cumul(x,y);
           if (cum>0) cimg_forC(*this,c) result(x,y,c) = (T)(blended(x,y,c)/cum);
         }
-      } else { cimg_forXY(blended,x,y) if (mask(x + ox,y + oy) && blend_map(x,y)==blend_iter) {
+      } else {
+        cimg_pragma_openmp(parallel for collapse(2) cimg_openmp_if(blended.width()>=256 && blended.height()>=16))
+        cimg_forXY(blended,x,y) if (mask(x + ox,y + oy) && blend_map(x,y)==blend_iter) {
           const float cum = cumul(x,y);
           if (cum>0) cimg_forC(*this,c) result(x,y,c) = (T)(blended(x,y,c)/cum);
         }
       }
     }
     if (is_blend_outer) draw_image(ox,oy,result);
-    else cimg_forXY(result,x,y) if (mask(x + ox,y + oy))
+    else {
+      cimg_pragma_openmp(parallel for collapse(2) cimg_openmp_if(result.width()>=256 && result.height()>=16))
+      cimg_forXY(result,x,y) if (mask(x + ox,y + oy))
            cimg_forC(*this,c) (*this)(x + ox,y + oy,c) = (T)result(x,y,c);
+    }
   }
   return *this;
 }
